### PR TITLE
Fix a bug for `CallbackHandler.callback_list`

### PR DESCRIPTION
**Pull request by harupy**
# What does this PR do?

Fix a bug where `CallbackHandler.callback_list` fails when given callbacks are duplicated:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-40-9605b122f4d1> in <module>()
      2 from transformers.trainer import DEFAULT_CALLBACKS
      3 
----> 4 CallbackHandler(DEFAULT_CALLBACKS + [MLflowCallback], "model", "optimizer", "lr_scheduler")

2 frames
/usr/local/lib/python3.6/dist-packages/transformers/trainer_callback.py in __init__(self, callbacks, model, optimizer, lr_scheduler)
    277         self.callbacks = []
    278         for cb in callbacks:
--> 279             self.add_callback(cb)
    280         self.model = model
    281         self.optimizer = optimizer

/usr/local/lib/python3.6/dist-packages/transformers/trainer_callback.py in add_callback(self, callback)
    299                 f"You are adding a {cb_class} to the callbacks of this Trainer, but there is already one. The current"
    300                 + "list of callbacks is\n:"
--> 301                 + self.callback_list
    302             )
    303         self.callbacks.append(cb)

/usr/local/lib/python3.6/dist-packages/transformers/trainer_callback.py in callback_list(self)
    326     @property
    327     def callback_list(self):
--> 328         return "\n".join(self.callbacks)
    329 
    330     def on_init_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl):

TypeError: sequence item 0: expected str instance, DefaultFlowCallback found
```

Code to reproduce the bug:

```python
from transformers.trainer_callback import CallbackHandler
from transformers.trainer import DEFAULT_CALLBACKS

CallbackHandler(DEFAULT_CALLBACKS + [MLflowCallback], "model", "optimizer", "lr_scheduler")
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sgugger

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by harupy** [comment]
Shoud I add a test verifying that instantiating a trainer with duplicated callbacks doesn't throw and emits a warning properly?

**Review by sgugger — APPROVED** [review]
Looking good, thanks for fixing! Any test you want to add is more than welcome as this new API is not super well tested yet.

**Comment by harupy** [comment]
@sgugger Thanks for the approval. I just added a test that verifies:

1. `Trainer` can be instantiated with duplicated callacks.
2. A warning is emitted for duplicated callbacks.

File: tests/test_trainer_callback.py [hunk 1]
```diff
@@ -221,3 +221,10 @@ def test_event_flow(self):
         trainer.train()
         events = trainer.callback_handler.callbacks[-2].events
         self.assertEqual(events, self.get_expected_events(trainer))
+
+        # warning should be emitted for duplicated callbacks
+        with unittest.mock.patch("transformers.trainer_callback.logger.warn") as warn_mock:
+            trainer = self.get_trainer(
+                callbacks=[MyTestTrainerCallback, MyTestTrainerCallback],
+            )
+            assert str(MyTestTrainerCallback) in warn_mock.call_args[0][0]
```
Comment (harupy) [inline tests/test_trainer_callback.py:230]: The warning message emitted here looks like:

```
You are adding a <class 'tests.test_trainer_callback.MyTestTrainerCallback'> to
the callbacks of this Trainer, but there is already one. The currentlist of callbacks is
:DefaultFlowCallback
TensorBoardCallback
MyTestTrainerCallback
```

Comment (harupy) [reply tests/test_trainer_callback.py:230]: `call_args[0][0]` corresponds to the first positional argument.

Comment (sgugger) [reply tests/test_trainer_callback.py:230]: Looks good!

**Review by sgugger — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]

**Pull request merged**